src/providers/cryptoauthlib/key_slot_storage.rs

```diff
@@ -0,0 +1,110 @@
+
+        // check (1)
+        match key_slots[key_id as usize].key_attr_vs_config(&key_attr) {
+            Ok(_) => (),
+            Err(err) => {
+                let error = std::format!("ATECC slot configuration mismatch: {}", err);
+                return Err(error);
+            }
+        };
+        // check(2)
+        match key_slots[key_id as usize].reference_check_and_set() {
+            Ok(_) => (),
+            Err(slot) => {
+                let warning = std::format!("Superfluous reference(s) to ATECC slot {:?}", slot);
+                return Ok(Some(warning));
+            }
+        };
+        match key_slots[key_id as usize].set_slot_status(KeySlotStatus::Busy) {
+            Ok(()) => Ok(None),
+            Err(err) => {
+                let error = std::format!("Unable to set hardware slot status: {}", err);
+                Err(error)
+            }
+        }
+    }
+
+    /// Lock protected per slot hardware configuration setter
+    pub fn set_hw_config(&self, hw_config: &[rust_cryptoauthlib::AtcaSlot]) -> Result<(), String> {
+        // RwLock protection
+        let mut key_slots = self.storage.write().unwrap();
+        for slot in 0..rust_cryptoauthlib::ATCA_ATECC_SLOTS_COUNT {
+            if hw_config[slot as usize].id != slot {
+                return Err(
+                    "configuration mismatch: vector index does not match its id.".to_string(),
+                );
+            }
+            key_slots[slot as usize] = AteccKeySlot {
+                ref_count: 0u8,
+                status: {
+                    match hw_config[slot as usize].is_locked {
+                        true => KeySlotStatus::Locked,
+                        _ => KeySlotStatus::Free,
+                    }
+                },
+                config: hw_config[slot as usize].config,
+            }
+        }
+        Ok(())
+    }
+
+    /// Lock protected set slot status wrapper
+    pub fn set_slot_status(
+        &self,
+        slot_id: usize,
+        status: KeySlotStatus,
+    ) -> Result<(), ResponseStatus> {
+        let mut key_slots = self.storage.write().unwrap();
+        key_slots[slot_id].set_slot_status(status)
+    }
+
+    /// Iterate through key_slots and find a free one with configuration matching attributes.
+    /// If found, the slot is marked Busy.
+    pub fn find_suitable_slot(&self, key_attr: &Attributes) -> Result<u8, ResponseStatus> {
+        let mut key_slots = self.storage.write().unwrap();
+        for slot in 0..rust_cryptoauthlib::ATCA_ATECC_SLOTS_COUNT {
+            if !key_slots[slot as usize].is_free() {
+                continue;
+            }
+            match key_slots[slot as usize].key_attr_vs_config(key_attr) {
+                Ok(_) => match key_slots[slot as usize].set_slot_status(KeySlotStatus::Busy) {
+                    Ok(()) => return Ok(slot),
+                    Err(err) => return Err(err),
```

Comment:
Could this be `=> continue,` as well, or is it critical to report the error?

Reply:
Difficult to say. This line is executed when there is an attempt to change slot status from "Locked" to "Busy". This is forbidden and this line should be unreachable. What to do when an unreachable branch is executed?

Reply:
There's always the [`unreachable!`](https://doc.rust-lang.org/std/macro.unreachable.html) macro! We've not used it much (because it panics if that path is ever taken), but it seems suitable here

Reply:
We have three policies to chose from
1. "ignore and go further", i.e. `=> continue`
2. "report the problem but try to continue". i.e. `=> return Err(err)`
3. "panic"

The first and the second give a chance to continue operation even though a problem with the logic is found. I would rather avoid calling panic (this way or another) in the field, but for tests it is fine. Maybe this would be the way to go? I mean something like this:
```
=> {
#[cfg(test)]
unreachable!();
#[cfg(not(test))]
return Err(err);
}
```

Reply:
It's ok, configuring it like that is completely overkill, was just curious about the initial reasoning about returning the error vs just continuing, I'm ok with things as they are